database/messages.js

```diff
@@ -1,69 +1,80 @@
 // connect to database
-require('./db');
+var db = require('./db');
 
 var mongoose = require('mongoose');
 var Promise = require('bluebird');
 mongoose.Promise = Promise;
 var messages = {};
 
 var MessageSchema = new mongoose.Schema({
+	id: Number,
 	text: String,
   	user: String,
   	reported: {type: Boolean, default: false},
   	date: { type: Date, default: Date.now }
 });
+MessageSchema.index({ id: 1});
 
 var MessageModel = mongoose.model('messages', MessageSchema);
 
-messages.saveMessage = function(data) {
-	var message = new MessageModel(
-	{
-		text: data.text,
-		user: data.user
+messages.saveMessage = function(data, callback) {
+	db.getNextAvailableId(db.messageIdKey, function(nextId) {
+		// create a new message object with given data and a unique id
+		var message = new MessageModel({
+			id: nextId,
+			text: data.text,
+			user: data.user
+		})
+		message.save();
+
+		db.setLastTakenId(db.messageIdKey, nextId, function(err) {
+			if (err) { console.log("error setting next id for messages: ", err); }
+        });
+
+        callback(message);
 	})
-	message.save();
 };
 
 messages.getReportedMessages = function(callback) {
-	var param = {reported: true};
-	MessageModel.find(param, function(err, data) {
-		console.log(data);
+	MessageModel.find({reported: true}, function(err, data) {
 		callback(data);
 	});
 }
 
-messages.report = function(user, text, callback) {
-	var param = {text: text, user: user}
-	MessageModel.update(param, {
+messages.report = function(messageID, callback) {
+	console.log(messageID);
+	MessageModel.update({id: messageID}, {
 		reported: true
-	}, function(){
-		callback();
-	});
+	}, callback);
 }
 
-messages.free = function(user, text, callback) {
-	var param = {text: text}; //, user: user}
-	MessageModel.update(param, {
+messages.free = function(messageID, callback) {
+	MessageModel.update({id: messageID}, {
 		reported: false
-	}, function(){
-		callback();
-	});
+	}, callback);
 }
 
-messages.delete = function(user, text, callback) {
+messages.delete = function(messageID, callback) {
 	//all distasteful messages regardless of user should be deleted
-	var param = {text: text};//, user: user};
-	MessageModel.remove(param, function(){
+	MessageModel.remove({id: messageID}, function(){
 		callback();
 	});
 }
 
-messages.next = function(id, volume, callback) {
-	var param = id != "" ? {_id:{"$lt": mongoose.Types.ObjectId(id)}} : null;
+messages.next = function(messageID, volume, callback) {
+	var param = messageID ? {id:{"$lt": messageID}} : null;
 	var promise = MessageModel.find(param).limit(parseInt(volume)).sort({"_id":-1}); 
 	promise.then(function(data){
 		callback(data);
 	})
 }
 
+messages.generateUsername = function() {
```

Comment:
Yeah, it's not worth the added complexity to prevent duplicate guest usernames

Remember, guest names are only relevant for messages as far back as users scroll. In order for someone to have a 1/10 chance of seeing another guest with the same username, they would have to scroll through 10,000 messages, assuming all unique visitors that only post one thing and then leave.. At that point, they would have been scrolling through messages for hours and React would long since have had performance issues

We have to remember that keeping our code organized and simple is very important, especially in an environment where people who wrote the code will be gone.